primitives/src/merkle_tree/field_based_mht/batch_mht/poseidon.rs

```diff
@@ -0,0 +1,491 @@
+        // postision until it has been processed relative to the subarray
+        let mut processed_pos_subarray = Vec::new();
+        // position until it has been processed relative to the full array
+        let mut processed_pos = Vec::new();
+        // position until new elements have been inserted in the subarray
+        let mut new_elem_pos_subarray = Vec::new();
+        // position until new elements have been inserted in the full array
+        let mut new_elem_pos = Vec::new();
+        // initial position of the full array
+        let mut initial_pos = Vec::new();
+
+
+        // keeps track of the Merkle tree level
+        let mut level_idx = 1;
+
+        let mut initial_idx_subarray = 0;
+        let mut initial_idx = 0;
+        let mut final_idx = last_level_size;
+
+        while size >= 1 {
+
+            // the size of sub-block was calculate above
+            let mut size_subblock = processing_block_size;
+            // if the size the of the sub-array that is part of the tree is greater than
+            // the corresponding real size, then use the real size to save memory space
+            if size_subblock > size {
+                size_subblock = size;
+            }
+
+            let final_idx_subarray = initial_idx_subarray + size_subblock;
+
+            final_pos.push(final_idx);
+            initial_pos_subarray.push(initial_idx_subarray);
+            final_pos_subarray.push(final_idx_subarray);
+            processed_pos.push(initial_idx);
+            processed_pos_subarray.push(initial_idx_subarray);
+            new_elem_pos.push(initial_idx);
+            new_elem_pos_subarray.push(initial_idx_subarray);
+            initial_pos.push(initial_idx);
+
+            initial_idx += size;
+            initial_idx_subarray += size_subblock;
+            size /= rate; // the next level of the tree will have leaves / rate
+            final_idx = initial_idx + size;
+            level_idx += 1;
+        }
+
+        let tree_size = *final_pos_subarray.last().unwrap();
+
+        let mut tree_nodes = Vec::with_capacity(tree_size);
+        for _i in 0..tree_size {
+            tree_nodes.push(F::zero());
+        }
+
+        Self {
+            root: { F::zero() },
+            array_nodes: { tree_nodes },
+            processing_step: { processing_block_size },
+            final_pos: { final_pos },
+            initial_pos: {initial_pos},
+            initial_pos_subarray: { initial_pos_subarray },
+            final_pos_subarray: { final_pos_subarray },
+            processed_pos: { processed_pos },
+            processed_pos_subarray: { processed_pos_subarray },
+            new_elem_pos: { new_elem_pos },
+            new_elem_pos_subarray: { new_elem_pos_subarray },
+            levels: { level_idx - 2 },
+            rate: { P::R },
+            num_leaves: { last_level_size },
+            _parameters: PhantomData
+        }
+    }
+
+    pub fn compute_subtree(&mut self) {
```

Comment:
No need to expose this function outside, let's keep it private (by removing the "pub")